cunumeric/module.py

```diff
@@ -6227,3 +6227,728 @@ def bincount(
+    pos = q * (n - 1)
+
+    # k = floor(pos)
+    # gamma = 1.0 if pos - k >= 0.5 else 0.0
+
+    k = np.round(pos)
+    gamma = 0.0
+
+    j = int(k)
+    return (gamma, j)
+
+
+# helper: test if array inputs are different:
+#
+def is_diff(arr1, arr2):
+    return len(arr1) != len(arr2) or any(
+        [x != y for (x, y) in zip(arr1, arr2)]
+    )
+
+
+# helper: in-place interpolator:
+# quantile interpolation method to act on vectors,
+# like an axpy call:
+# [gamma](vleft, vright) { return (1-gamma)*vleft + gamma*vright;}
+#
+def in_place_interp(gamma, arr_lvals, arr_rvals, result):
+    result[...] = (1.0 - gamma) * arr_lvals + gamma * arr_rvals
+
+
+# for the case when axis = tuple (non-singleton)
+# reshuffling might have to be done (if tuple is non-consecutive)
+# and the src array must be collapsed along that set of axes
+#
+# args:
+#
+# arr:    [in] source nd-array on which quantiles are calculated;
+# axes_set: [in] tuple or list of axes (indices less than arr dimension);
+#
+# return: pair: (minimal_index, reshuffled_and_collapsed source array)
+# TODO: check if reshuffling, reshaping is done in-place!
+#
+def reshuffle_reshape(arr, axes_set):
+    ndim = len(arr.shape)
+
+    sorted_axes = tuple(sorted(axes_set))
+
+    min_dim_index = sorted_axes[0]
+    num_axes = len(sorted_axes)
+    reshuffled_axes = tuple(range(min_dim_index, min_dim_index + num_axes))
+
+    non_consecutive = is_diff(sorted_axes, reshuffled_axes)
+    if non_consecutive:
+        arr_shuffled = moveaxis(arr, sorted_axes, reshuffled_axes)
+    else:
+        arr_shuffled = arr
+
+    # shape_reshuffled = arr_shuffled.shape # debug
+    collapsed_shape = np.product(
+        [arr_shuffled.shape[i] for i in reshuffled_axes]
+    )
+
+    redimed = tuple(range(0, min_dim_index + 1)) + tuple(
+        range(min_dim_index + num_axes, ndim)
+    )
+    reshaped = tuple(
+        [
+            collapsed_shape if k == min_dim_index else arr_shuffled.shape[k]
+            for k in redimed
+        ]
+    )
+
+    arr_reshaped = arr_shuffled.reshape(reshaped)
+    return (min_dim_index, arr_reshaped)
+
+
+# args:
+#
+# arr:      [in] source nd-array on which quantiles are calculated;
+#                preccondition: assumed sorted!
+# q_arr:    [in] quantile input values nd-array;
+# axis:     [in] axis along which quantiles are calculated;
+# method:   [in] func(q, n) returning (gamma, j),
+#                where = array1D.size;
+# keepdims: [in] boolean flag specifying whether collapsed axis
+#                should be kept as dim=1;
+# to_dtype: [in] dtype to convert the result to;
+# qs_all:   [in/out] result pass through or created (returned)
+#
+def quantile_impl(
+    arr,
+    q_arr,
+    axis,
+    axes_set,
+    original_shape,
+    method,
+    keepdims,
+    to_dtype,
+    qs_all,
+):
+
+    ndims = len(arr.shape)
+
+    if axis is None:
+        n = arr.size
+
+        if keepdims:
+            remaining_shape = tuple(ones(len(axes_set), dtype=int))
```

Comment:
No need to go through a cuNumeric array, then back to a tuple. You can make a tuple of 1s directly with:
```
remaining_shape = (1,) * len(axes_set)
```

Reply:
Done.